modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+/**
+ * Copyright (C) 2011 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.basics.currency;
+
+import static com.opengamma.collect.Guavate.entriesToImmutableMap;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.opengamma.collect.ArgChecker;
+import com.opengamma.collect.tuple.Pair;
+
+/**
+ * Immutable class describing a set of currencies and all the cross rates between them.
+ */
+public class FxMatrix {
+
+  /**
+   * An FX matrix containing neither currencies nor rates.
+   */
+  public static final FxMatrix EMPTY_FX_MATRIX = builder().build();
+
+  /**
+   * The map between the currencies and their order. An ImmutableMap is
+   * used so that the currencies are correctly ordered when the
+   * {@link #toString()} method is called.
+   */
+  private final ImmutableMap<Currency, Integer> currencies;
+
+  /**
```

Comment:
The field names in the example code in this comment have underscore prefixes like the 2.x platform but the actual variables in the class don't. Is that intentional?